newrelic/metric.go

```diff
@@ -0,0 +1,226 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package newrelic
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/google/ts-bridge/record"
+
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/appengine/log"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+// MetricConfig defines configuration file parameters for a specific metric imported from New Relic.
+type MetricConfig struct {
+	APIKey        string `yaml:"api_key" validate:"nonzero"`
+	ApplicationId string `yaml:"application_id" validate:"nonzero"`
+	MetricName    string `validate:"nonzero" yaml:"metric_name"`
+	MetricValue   string `validate:"nonzero" yaml:"metric_value"`
+	Cumulative    bool
+}
+
+// Metric defines a New Relic-based metric. It implements the SourceMetric interface.
+type Metric struct {
+	Name   string
+	config *MetricConfig
+	client *Client
+}
+
+type timeslice struct {
+	From   string             `json:"from"`
+	To     string             `json:"to"`
+	Values map[string]float64 `json:"values"`
+}
+type data struct {
+	Name       string      `json:"name"`
+	Timeslices []timeslice `json:"timeslices"`
+}
+type metricData struct {
+	From            string   `json:"from"`
+	To              string   `json:"to"`
+	MetricsNotFound []string `json:"metrics_not_found,omitempty"`
+	MetricsFound    []string `json:"metrics_found,omitempty"`
+	Metrics         []data   `json:"metrics`
+}
+type responseData struct {
+	MetricData metricData `json:"metric_data"`
+}
+
+// Client defines the information necessary to query New Relic
+type Client struct {
+	APIKey        string
+	ApplicationId string
+	BaseUrl       string
+}
+
+// Query returns a unique name for the query
+func (m *Metric) Query() string {
+	return fmt.Sprintf("%v:%v", m.config.MetricName, m.config.MetricValue)
+}
+
+// QueryMetrics returns a metric descriptor and a slice of TimeSeries for the metric.
+// all elements of the TimeSeries must be the same metric descriptor and contain a single point
+func (m *Metric) QueryMetrics(ctx context.Context, start time.Time, end time.Time, name string, value string) (*metricpb.MetricDescriptor, []*monitoringpb.TimeSeries, error) {
+	c := m.client
+
+	var ret []*monitoringpb.TimeSeries
+	tmpStart := start
+
+	for continueLoop := true; continueLoop; {
+		// Always use 3 hour windows, anything more and points will get aggregated to greater than 1 minute granularity
+		tmpEnd := tmpStart.Add(3 * time.Hour)
+		if tmpEnd.After(end) {
+			tmpEnd = end
+			continueLoop = false
+		}
+
+		// TODO: add multi-page support
```

Comment:
with a 3 hour window, I'm pretty sure that the max is 180 results and they document that they start paginating at 200 results, so I don't think it's possible to hit this, but it is possible. My thinking was that if we did get paged data, we would not process it and it would just be included in the next sync, right?

Reply:
API documentation says that paged data [is returned in an indeterminate order](https://docs.newrelic.com/docs/apis/rest-api-v2/requirements/pagination-api-output) (though I am not sure whether this is the case in practice), so we might be missing some points if we blindly process the first page of the response.

If we never expect paginated responses, I'd prefer if we treated them as errors instead of processing page 1 as if it was complete result.